lib/pbench/server/database/models/users.py

```diff
@@ -1,58 +1,121 @@
+    """
+
+
+class UserSqlError(UserError):
+    """SQLAlchemy errors reported through User operations.
+
+    The exception will identify the operation being performed and the config
+    key; the cause will specify the original SQLAlchemy exception.
+    """
+
+    def __init__(self, operation: str, params: JSONOBJECT, cause: str):
+        self.operation = operation
+        self.params = params
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Error {self.operation} {self.params!r}: {self.cause}"
+
+
+class UserDuplicate(UserError):
+    """Attempt to commit a duplicate unique value."""
+
+    def __init__(self, user: "User", cause: str):
+        self.user = user
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Duplicate user setting in {self.user.get_json()}: {self.cause}"
+
+
+class UserNullKey(UserError):
+    """Attempt to commit an Audit row with an empty required column."""
+
+    def __init__(self, user: "User", cause: str):
+        self.user = user
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Missing required key in {self.user.get_json()}: {self.cause}"
+
+
 class User(Database.Base):
     """User Model for storing user related details."""
 
     __tablename__ = "users"
 
     id = Column(Integer, primary_key=True, autoincrement=True)
-    username = Column(String(255), unique=True, nullable=False)
-    first_name = Column(String(255), unique=False, nullable=False)
-    last_name = Column(String(255), unique=False, nullable=False)
-    password = Column(LargeBinary(128), nullable=False)
-    registered_on = Column(DateTime, nullable=False, default=datetime.datetime.now())
-    email = Column(String(255), unique=True, nullable=False)
-    role = Column(Enum(Roles), unique=False, nullable=True)
-    auth_tokens = relationship("ActiveTokens", backref="users")
+    oidc_id = Column(String(255), unique=True, nullable=False)
+    username = Column(String(255), unique=True, nullable=True)
+    dataset_metadata = relationship(
+        "Metadata", back_populates="user", cascade="all, delete-orphan"
+    )
+    _roles = Column(String(255), unique=False, nullable=True)
+
+    @property
+    def roles(self):
+        if self._roles:
+            return [x for x in self._roles.split(";")]
+        else:
+            return []
+
+    @roles.setter
+    def roles(self, value):
+        if isinstance(value, list):
+            self._roles = ";".join(value)
+        else:
+            raise UserSqlError("Adding role", value, "Value is not a list")
 
     def __str__(self):
         return f"User, id: {self.id}, username: {self.username}"
 
     def get_json(self):
-        return {
-            "username": self.username,
-            "email": self.email,
-            "first_name": self.first_name,
-            "last_name": self.last_name,
-            "registered_on": self.registered_on,
-        }
+        return {"username": self.username, "id": self.id}
```

Comment:
Return `username` and `oidc_id` here, not the internal and meaningless `id`. You're only using this for an ADMIN asking about some user, and they specified the `username` to the `GET` API; if all they wanted was the `username` they wasted a call!